pkg/resources/management.cattle.io/v3/roletemplate/main_test.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/rancher/webhook/pkg/admission"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+	v1 "k8s.io/api/admission/v1"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+var errTest = errors.New("bad error")
+
+type tableTest struct {
+	wantRT    func() *v3.RoleTemplate
+	name      string
+	args      args
+	wantError bool
+	allowed   bool
+}
+
+type args struct {
+	oldRT    func() *v3.RoleTemplate
+	newRT    func() *v3.RoleTemplate
+	username string
+}
+type RoleTemplateSuite struct {
+	suite.Suite
+	ruleEmptyVerbs rbacv1.PolicyRule
+	adminRT        *v3.RoleTemplate
+	readNodesRT    *v3.RoleTemplate
+	lockedRT       *v3.RoleTemplate
+	adminCR        *rbacv1.ClusterRole
+	manageNodeRole *rbacv1.ClusterRole
+}
+
+func TestRoleTemplates(t *testing.T) {
+	t.Parallel()
+	suite.Run(t, new(RoleTemplateSuite))
+}
+
+func (c *RoleTemplateSuite) SetupSuite() {
+	ruleReadPods := rbacv1.PolicyRule{
+		Verbs:     []string{"GET", "WATCH"},
+		APIGroups: []string{"v1"},
+		Resources: []string{"pods"},
+	}
+	ruleWriteNodes := rbacv1.PolicyRule{
+		Verbs:     []string{"PUT", "CREATE", "UPDATE"},
+		APIGroups: []string{"v1"},
+		Resources: []string{"nodes"},
+	}
+	ruleAdmin := rbacv1.PolicyRule{
+		Verbs:     []string{"*"},
+		APIGroups: []string{"*"},
+		Resources: []string{"*"},
+	}
+	c.ruleEmptyVerbs = rbacv1.PolicyRule{
+		Verbs:     nil,
+		APIGroups: []string{"v1"},
+		Resources: []string{"nodes"},
+	}
+	c.readNodesRT = &v3.RoleTemplate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "read-role",
+		},
+		DisplayName: "Read Role",
+		Rules:       []rbacv1.PolicyRule{ruleReadPods, ruleWriteNodes},
+		Context:     "cluster",
+	}
+	c.adminRT = &v3.RoleTemplate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "admin-role",
+		},
+		DisplayName:    "Admin Role",
+		Rules:          []rbacv1.PolicyRule{ruleAdmin},
+		Builtin:        true,
+		Administrative: true,
+		Context:        "cluster",
+	}
+	c.lockedRT = &v3.RoleTemplate{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "locked-role",
+		},
+		DisplayName: "Locked Role",
+		Rules:       []rbacv1.PolicyRule{ruleReadPods},
+		Locked:      true,
+		Context:     "cluster",
+	}
+	c.adminCR = &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "admin-role",
+		},
+		Rules: []rbacv1.PolicyRule{ruleAdmin},
+	}
+	c.manageNodeRole = &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{Name: "manage-nodes"},
+		Rules:      []rbacv1.PolicyRule{ruleReadPods, ruleWriteNodes},
+	}
+}
+
+// createRTRequest will return a new webhookRequest with the using the given RTs
+// if oldRT is nil then a request will be returned as a create operation.
+// if newRT is nil then a request will be returned as a delete operation.
+// else the request will look like and update operation.
```

Comment:
I think this is what you meant:
```suggestion
// else the request will look like an update operation.
```